lib/ansible/modules/cloud/amazon/s3_bucket.py

```diff
@@ -250,13 +208,81 @@ def _create_or_update_bucket(connection, module, location):
                 current_tags_dict = tags
                 changed = True
             except S3ResponseError as e:
-                module.fail_json(msg=e.message)
+                module.fail_json(msg='Failed to set tags for S3 bucket: %s' % e)
+
+    return current_tags_dict
+
+
+def create_or_update_bucket(connection, module, location, flavour='aws'):
+    name = module.params.get("name")
+    requester_pays = module.params.get("requester_pays")
+    versioning = module.params.get("versioning")
+    changed = False
+
+    try:
+        bucket = connection.get_bucket(name)
+    except S3ResponseError as e:
+        try:
+            bucket = connection.create_bucket(name, location=location)
+            changed = True
+        except S3CreateError as e:
+            module.fail_json(msg='Failed to create S3 bucket: %s' % e)
+
+    if not bucket:
+        module.fail_json(msg='Unable to create bucket, no error from the API')
+
+    # Versioning
+    versioning_status = bucket.get_versioning_status()
+    if versioning is not None:
+        if versioning and versioning_status.get('Versioning') != "Enabled":
+            try:
+                bucket.configure_versioning(versioning)
+                changed = True
+                versioning_status = bucket.get_versioning_status()
+            except S3ResponseError as e:
+                module.fail_json(msg='Failed to enable versioning for S3 bucket: %s' % e, exception=traceback.format_exc())
+        elif not versioning and versioning_status.get('Versioning') == "Enabled":
+            try:
+                bucket.configure_versioning(versioning)
+                changed = True
+                versioning_status = bucket.get_versioning_status()
+            except S3ResponseError as e:
+                module.fail_json(msg='Failed to disable versioning for S3 bucket: %s' % e, exception=traceback.format_exc())
+
+    # Requester pays
+    requester_pays_status = get_request_payment_status(bucket)
+    if requester_pays_status != requester_pays:
+        if requester_pays:
+            payer='Requester'
+        else:
+            payer='BucketOwner'
+        bucket.set_request_payment(payer=payer)
+        changed = True
+        requester_pays_status = get_request_payment_status(bucket)
+
+    # Policy
+    current_policy=None
+    if flavour != 'ceph':
+        current_policy = _update_policy(module, bucket)
+    if module.params["policy"] is not None and flavour == 'ceph':
+        # policy not supported by Ceph Object Gateway (radosgw)
+        # http://docs.ceph.com/docs/master/radosgw/s3/#features-support
+        module.fail_json(msg="policy not supported by ceph object gateway")
+
+    # Tags
+    current_tags_dict=None
+    if flavour != 'ceph':
+        current_tags_dict = _update_tags(module, bucket)
+    if module.params["tags"] is not None and flavour == 'ceph':
```

Comment:
Maybe make this an `elif module.params["tags"] is not None:` since the `== 'ceph'` is the inverse of above?